File: tools/test_websockets.py
#!/usr/bin/env python3
"""
WebSocket Test Client
=====================

This script connects to the FastAPI WebSocket endpoints for both logs and state
to display the real-time data being broadcasted.

Usage:
    python scripts/test_websockets.py

You must have the main `dashboard_api.py` running in another terminal for this
script to connect successfully.
"""
import asyncio
import json
import websockets

async def listen_to_endpoint(uri, name):
    """A generic function to connect to a WebSocket endpoint and print messages."""
    while True:
        try:
            async with websockets.connect(uri) as websocket:
                print(f"[{name}] Successfully connected to {uri}")
                while True:
                    message = await websocket.recv()
                    try:
                        # Pretty-print JSON if possible
                        data = json.loads(message)
                        print(f"[{name}] Received:")
                        print(json.dumps(data, indent=2))
                    except json.JSONDecodeError:
                        print(f"[{name}] Received (raw): {message}")
        except (websockets.exceptions.ConnectionClosed, ConnectionRefusedError) as e:
            print(f"[{name}] Connection to {uri} lost: {e}. Retrying in 5 seconds...")
            await asyncio.sleep(5)
        except Exception as e:
            print(f"[{name}] An unexpected error occurred: {e}. Retrying in 5 seconds...")
            await asyncio.sleep(5)


async def main():
    """Runs the two listeners concurrently."""
    logs_uri = "ws://127.0.0.1:8000/ws/logs"
    state_uri = "ws://127.0.0.1:8000/ws/state"
    
    print("--- Starting WebSocket Test Client ---")
    print(f"Attempting to connect to:")
    print(f"  - Logs: {logs_uri}")
    print(f"  - State: {state_uri}")
    print("------------------------------------")

    log_listener = listen_to_endpoint(logs_uri, "LOGS")
    state_listener = listen_to_endpoint(state_uri, "STATE")

    await asyncio.gather(
        log_listener,
        state_listener
    )

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\nClient stopped by user.") 

File: tools/test_websockets.js

import WebSocket from 'ws';

const listenToEndpoint = (uri, name, messageLimit, onDisconnect) => {
    const ws = new WebSocket(uri);
    let messageCount = 0;

    // Add timeout to wait longer for messages
    const timeout = setTimeout(() => {
        console.log(`[${name}] Timeout reached. Closing connection.`);
        ws.close();
    }, 10000); // 10 second timeout

    ws.on('open', () => {
        console.log(`[${name}] Successfully connected to ${uri}`);
    });

    ws.on('message', (message) => {
        if (messageCount >= messageLimit) return;

        try {
            const data = JSON.parse(message);
            console.log(`[${name}] Received:`);
            console.log(JSON.stringify(data, null, 2));
        } catch (error) {
            console.log(`[${name}] Received (raw): ${message}`);
        }

        messageCount++;
        if (messageCount >= messageLimit) {
            console.log(`[${name}] Received ${messageLimit} messages. Closing connection.`);
            clearTimeout(timeout);
            ws.close();
        }
    });

    ws.on('error', (error) => {
        console.error(`[${name}] An unexpected error occurred: ${error.message}.`);
        clearTimeout(timeout);
        onDisconnect();
    });

    ws.on('close', () => {
        console.log(`[${name}] Connection to ${uri} closed.`);
        clearTimeout(timeout);
        onDisconnect();
    });
};

const main = () => {
    const logsUri = 'ws://127.0.0.1:8000/ws/logs';
    const stateUri = 'ws://127.0.0.1:8000/ws/state';
    const messageLimit = 3; // Increased to 3 messages
    let openConnections = 2;

    const handleDisconnect = () => {
        openConnections--;
        if (openConnections === 0) {
            console.log('All WebSocket connections closed. Exiting.');
            process.exit(0);
        }
    };

    console.log('--- Starting WebSocket Test Client ---');
    console.log(`Fetching ${messageLimit} messages from each endpoint.`);
    console.log('Attempting to connect to:');
    console.log(`  - Logs: ${logsUri}`);
    console.log(`  - State: ${stateUri}`);
    console.log('------------------------------------');

    listenToEndpoint(logsUri, 'LOGS', messageLimit, handleDisconnect);
    listenToEndpoint(stateUri, 'STATE', messageLimit, handleDisconnect);
};

main();
